src/pages/Groups.tsx

```diff
@@ -0,0 +1,253 @@
+        undefined
+    )
+
+    const [isModalShowing, setIsModalShowing] = useState(false)
+
+    const [modalGroup, setModalGroup] = useState<Group | undefined>(undefined)
+
+    //tracking first render
+    const isMounted = useMounted()
+
+    /**
+     * Api search request side effect.
+     *
+     * How this works (general ideas):
+     * 1) gets called every time search text changes or rescheduleSearch is incremented (for forcing search request)
+     * 2) if len(search) < 5 do nothing and delete stored groups if there are any.
+     * 3) if no other searches were done before, send search request to api. Store time of search in `lastSearchTime`.
+     * 4) successive requests are allowed if last search request time (successful or not) was more than `deltaTime` (500ms) ago
+     * 5) if a request is not allowed, store `lastTimeSearch` anyway, then check `deltaTime` (500ms) later if the user has kept writing.
+     * if he has, then do nothing because side effect will be called again, if he hasn't,
+     * then force search request with latest search value.
+     */
+    const searchGroups = async () => {
+        if (isMounted) {
+            if (search.length < 5) {
+                if (groups !== undefined) {
+                    setGroups(undefined)
+                }
+                return
+            }
+            try {
+                const now = new Date()
+                const msPassed = msPassedBetween(lastSearchTime, now)
+                if (
+                    msPassed === undefined ||
+                    msPassed >= deltaTime ||
+                    needSearching === true
+                ) {
+                    //update last time search
+                    setLastSearchTime(now)
+                    const response = await api.groups.get(
+                        {
+                            name: search.trimEnd(),
+                            year: year === all ? undefined : year,
+                            degree: course === all ? undefined : course,
+                            type: type === all ? undefined : type,
+                            platform: platform === all ? undefined : platform,
+                        },
+                        { maxRetries: 1, retryType: RetryType.RETRY_N_TIMES }
+                    )
+                    setGroups(response)
+                    //reset need searching for next render
+                    setNeedSearching(false)
+                } else {
+                    setLastSearchTime(now)
+                    await wait(deltaTime)
+                    //in the next render you have the previous and current search text, and can compare changes
+                    //this triggers Reschedule Search Side Effect
+                    setPrevSearch(search)
+                }
+            } catch (error) {
+                console.log(error)
+            }
+        }
+    }
+    useEffect(() => {
+        void searchGroups()
+    }, [search, rescheduleSearch])
```

Comment:
se vuoi limitare le chiamate fatte in un deltatime, basta cancellare dei timeout
```suggestion
    useEffect(() => {
        clearTimeout(searchTimeout)
        searchTimeout = setTimeout(() => {
            void searchGroups()
        }, deltaTime)
    }, [search])
```
con un 
```ts
let searchTimeout: NodeJS.Timeout
```
dichiarato in cima al file fuori dal componente

Reply:
Fantastico!